android/src/main/java/com/pspdfkit/react/menu/ReactGroupingRule.java

```diff
@@ -0,0 +1,117 @@
+package com.pspdfkit.react.menu;
+
+import android.content.Context;
+import android.support.annotation.IdRes;
+import android.support.annotation.IntRange;
+import android.support.annotation.NonNull;
+
+import com.facebook.react.bridge.Dynamic;
+import com.facebook.react.bridge.ReadableArray;
+import com.facebook.react.bridge.ReadableMap;
+import com.facebook.react.bridge.ReadableType;
+import com.pspdfkit.react.R;
+import com.pspdfkit.ui.toolbar.ContextualToolbar;
+import com.pspdfkit.ui.toolbar.grouping.presets.MenuItem;
+import com.pspdfkit.ui.toolbar.grouping.presets.PresetMenuItemGroupingRule;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ReactGroupingRule extends PresetMenuItemGroupingRule {
+
+    private static final int INVALID_ID = -1;
+
+    private final List<MenuItem> items = new ArrayList<>();
+
+    public ReactGroupingRule(@NonNull Context context, @NonNull ReadableArray menuItems) {
+        super(context);
+        for (int i = 0; i < menuItems.size(); i++) {
+            Dynamic menuItem = menuItems.getDynamic(i);
+            if (menuItem.getType() == ReadableType.Map) {
+                ReadableMap groupItem = menuItem.asMap();
+                int key = getIdFromName(groupItem.getString("key"));
+                if (key == INVALID_ID) {
+                    continue;
+                }
+                ReadableArray subitems = groupItem.getArray("items");
+                int[] itemIds = new int[subitems.size()];
+                for (int j = 0; j < itemIds.length; j++) {
```

Comment:
Can we extract inner `for` along with few more related lines (starting with `ReadableArray subitems = g...`)and give it a good name like `collectItemIdsFormName`for easier reading?